R/textmodel_nb.R

```diff
@@ -331,3 +331,80 @@ colNorm <- function(x) {
+#' @examples 
+#' txt <- c(d1 = "Chinese Beijing Chinese",
+#'          d2 = "Chinese Chinese Shanghai",
+#'          d3 = "Chinese Macao",
+#'          d4 = "Tokyo Japan Chinese",
+#'          d5 = "Chinese Chinese Chinese Tokyo Japan")
+#' trainingset <- dfm(txt, tolower = FALSE)
+#' trainingclass <- factor(c("Y", "Y", "Y", "N", NA), ordered = TRUE)
+#' tmod <- textmodel_nb(trainingset, y = trainingclass, prior = "docfreq")
+#' gain.textmodel_nb(tmod)
+#' gain.textmodel_nb(tmod, class_specific = FALSE)
+#' @export
+gain.textmodel_nb <- function(model, base = 2, class_specific = TRUE) {
+    if (!("textmodel_nb" %in% class(model))) stop("model must be a textmodel_nb object")
+    if (class(class_specific) != "logical") stop("class_specific must be a logical (TRUE or FALSE)")
+
+    # The formula for entropy of word w is
+    # - summation over all classes ( P(class) * log P(class))  (Term 1)
+    # + P(w) * summation over all classes (P(class |w ) * log(P(class|w)) ) (Term 2)
+    # + P(w') * summation over all classes (P(class | w') * log(P(class|w')) ) (Term 3)
+
+    # Term 1: Initial Entropy
+    # Get summation over all classes of  P(class) * log P(class)
+    initial_entropy <- apply(data.frame(model$Pc), MARGIN = 2, function(x) -1 * x * log(x, base = base))
+    if (class_specific == F) initial_entropy <- sum(initial_entropy)
+
+    # Term 2: Entropy Given w
+    # Get the conditional entropy := summation over all classes P(class|w) * log(P(class|w))
+    conditional_entropy <- apply(model$PcGw, MARGIN = 2, function(x) x * log(x, base = base))
+    if (class_specific == F) conditional_entropy <- colSums(conditional_entropy)
+
+    # Now get the unconditional entropy  := P(w) * conditional entropy
+    unconditional_entropy <-  t(data.frame(t(conditional_entropy)) * model$Pw)
+
+    # Term 3: Entropy Given w'
+    # Get the second conditional entropy := summation over all classes P(class | w') * log(P(class|w'))
+    # where P(w') denotes the probability that a word is not in a document
+    # To derive P(class|w'), we note:
+    # P(class | w') = P(class ∩ w') / P(w') = (P(class) - p(class ∩ w)) / P(w')
+    # = ( P(class) - [P(w|c) * P(c)] ) / ( 1 - P(w')) # Bayes Rule
+    Pcintersectw <- data.frame(model$PwGc) * model$Pc
+    Pcintersectwcomp <- (model$Pc - Pcintersectw)
+    PcGwcomp <- t(data.frame(t(Pcintersectwcomp)) / (1-model$Pw)) # P(class | w') =  P(class ∩ w') / P(w')
+    # PcGwcomp is a dataframe that captures conditional on a word not being in the document,
+    # what is the probability that it belongs to a class?
+
+    conditional_entropy_comp <- apply(PcGwcomp, MARGIN = 2, function(x) x * log(x, base = base))
+    if (class_specific == F) conditional_entropy_comp <- colSums(conditional_entropy_comp)
+
+    unconditional_entropy_comp <- t((1 - model$Pw) * data.frame(t(conditional_entropy_comp)))
+
+    # Add the three terms and return either as matrix of named-list
+    if (class_specific == T) {
```

Comment:
if this is logical then you can omit the `== TRUE` and just use `else` for the `FALSE` case